calc/active_rules_calculator.go

```diff
@@ -282,3 +281,47 @@ func (arc *ActiveRulesCalculator) sendPolicyUpdate(policyKey model.PolicyKey) {
 		arc.RuleScanner.OnPolicyInactive(policyKey)
 	}
 }
+
+// EndpointKeyToProfileIDMap is a specialised map that calculates the deltas to the profile IDs
+// when making an update.
+type EndpointKeyToProfileIDMap struct {
+	endpointKeyToProfileIDs map[model.Key][]string
+}
+
+func NewEndpointKeyToProfileIDMap() *EndpointKeyToProfileIDMap {
+	return &EndpointKeyToProfileIDMap{
+		endpointKeyToProfileIDs: make(map[model.Key][]string),
+	}
+}
+
+func (idx EndpointKeyToProfileIDMap) Update(
+	key model.Key,
+	profileIDs []string,
+) (
+	removedIDs, addedIDs map[string]bool,
+) {
+	oldIDs := idx.endpointKeyToProfileIDs[key]
```

Comment:
If key isn't in the map, oldIDs will be a nil slice, which is safe to iterate over (it simply executes the loop 0 times).  It's also safe to delete something that's not in the map.